styx-api-service/src/main/java/com/spotify/styx/api/WorkflowResource.java

```diff
@@ -219,31 +217,32 @@ private WorkflowConfiguration readFromJsonWithDefaults(ByteString payload)
     return WorkflowConfigurationBuilder.from(workflowConfig).deploymentTime(time.get()).build();
   }
 
-  private Response<Collection<WorkflowWithState>> workflowsWithState() {
-    try {
-      Collection<WorkflowWithState> workflows = storage.workflowsWithState().values();
-      return Response.forPayload(workflows);
-    } catch (IOException e) {
-      throw new RuntimeException("Failed to get workflows", e);
-    }
-  }
-
-  private Response<Collection<Workflow>> workflows(Request request) {
+  private Response<? extends Collection<?>> workflows(Request request) {
     try {
       var paramFilters = Stream.of(QueryParams.values())
           .map(e -> getFilterParams(request, e).map(param -> Map.entry(e, param)))
           .flatMap(Optional::stream)
           .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
 
-      Collection<Workflow> workflows = storage.workflows().values();
+      Collection<WorkflowWithState> workflowsWithState = storage.workflowsWithState().values();
+      Collection<WorkflowWithState> filteredWorkflows = filterWorkflows(workflowsWithState, paramFilters);
 
-      return Response.forPayload(filterWorkflows(workflows, paramFilters));
+      if (includeStates(request)) {
+        return Response.forPayload(filteredWorkflows);
+      }
 
+      List<Workflow> workflows = filteredWorkflows.stream().map(WorkflowWithState::workflow).collect(Collectors.toList());
+      return Response.forPayload(workflows);
     } catch (IOException e) {
       throw new RuntimeException("Failed to get workflows", e);
     }
   }
 
+  private boolean includeStates(Request request) {
+    Optional<String> full = getFilterParams(request, INCLUDE_STATES);
```

Comment:
`getFilterParams(request, INCLUDE_STATES).isPresent()` should work without checking the actual value.